### ISSUE TITLE
[APPIUM 1.5.0] Frequent random "This socket has been ended by the other party"

### ISSUE THREAD

**Issue by mittt** [appium/appium#6332]
## The problem

I have 4 tests running in sequence and i get the following error at random. The same test may pass successfully next time I start them all.

**On the client**
Original error: Android boostrap socket crashed: Error: This socket has been ended by the other party  

**In the Appium log**
2016-03-24 13:58:47:421 - error: [MJSONWP] Encountered internal error running command: Error: Android boostrap socket crashed: Error: This socket has been ended by the other party
    at Socket.<anonymous> (lib/bootstrap.js:87:21)
    at emitOne (events.js:77:13)

## Environment

* Appium version (or git revision) that exhibits the issue: └── appium@1.5.0
* Desktop OS/version used to run Appium: Windows 10
* Node.js version (unless using Appium.app|exe): node: '5.0.0',
* Mobile platform/version under test: Android, 
* Real device or emulator/simulator: 5 devices, 5 instances of appium running connected to a selenium hub, 5 tests running in parallel on those devices
* Appium CLI or Appium.app|exe: CLI

## Link to Appium logs

[On the Client](https://gist.github.com/mittt/a32b0f67d6d9057c08ce)
[In Appium](https://gist.github.com/mittt/c8ff9f869ff1af177647)

**Comment by saikrishna321** [comment]
@mittt this has been fixed in the source with "appium-android-bootstrap": "^2.5.2"

https://github.com/appium/appium/issues/6202

**Comment by riverole** [comment]
Sorry to bother here, but, I am facing this issue. I installed appium through npm (the simple npm i -g appium), and I'm having a hard time to update the version of appium-android-bootstrap to 2.5.2 (I have currently 2.5.1). I have seen that the normal npm update only works for top packages, which in this case is appium, and I tried npm -g --depth Infinity update, but it doesn't seem to do the trick either. I have basically 0 idea about node and npm, could someone point me to how to achieve this? Thanks a lot!

**Comment by saikrishna321** [comment]
@riverole you need to clone the appium code and install from source 

https://github.com/appium/appium/blob/master/docs/en/contributing-to-appium/appium-from-source.md

**Comment by mittt** [comment]
Is this error also related:
Session [(null externalkey)] not available and is not among the last 1000 terminated sessions.

I'm getting it the same way - Frequent, but random.

**Comment by imurchie** [comment]
Appium uses an shrinkwrap file to lock in dependencies, so a normal install won't get the latest.

You can use the `--no-shrinkwrap` flag to override.

```
npm install -g appium --no-shrinkwrap
```

**Comment by riverole** [comment]
thanks @imurchie, that actually made the trick! I tried also your option @saikrishna321 and it worked too, but I think this way (with the no-shrinkwrap option) it will be less disruptive on the way we use appium in our scripts right now. Thanks a lot to both of you!

**Comment by imurchie** [comment]
You're welcome.

**Comment by zqy2084** [comment]
@imurchie I have updated appium to 1.5.1, I still got the same problem:
├─┬ appium@1.5.1
│ ├─┬ appium-android-driver@1.8.1
│ │ ├── appium-adb@2.5.1
│ │ ├─┬ appium-android-bootstrap@2.6.0

Is the fix included in the latest 1.5.1?

**Comment by saikrishna321** [comment]
@zqy2084 yup this has been fixed and updated to 1.5.1 

Can you please post your server logs ?

**Comment by zqy2084** [comment]
@saikrishna321 Found it is not exactly the same error output, so i file a new issue as:
https://github.com/appium/appium/issues/6362
